src/main/java/org/spongepowered/common/resource/SpongeResource.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.resource;
+
+import net.minecraft.server.packs.resources.Resource;
+import org.spongepowered.api.ResourceKey;
+import org.spongepowered.api.resource.ResourcePath;
+
+import java.io.InputStream;
+
+public final class SpongeResource implements org.spongepowered.api.resource.Resource {
+
+    private final ResourcePath path;
+    private InputStream stream;
+
+    public SpongeResource(final Resource resource) {
+        this.path = new SpongeResourcePath((ResourceKey) (Object) resource.getLocation());
+        this.stream = resource.getInputStream();
+    }
+
+    public SpongeResource(final ResourcePath path, final InputStream stream) {
+        this.path = path;
+        this.stream = stream;
+
+    }
+
+    @Override
+    public ResourcePath path() {
+        return this.path;
+    }
+
+    @Override
+    public InputStream inputStream() {
+        if (this.stream == null) {
```

Comment:
So, this still has the inherent issue of people not closing the stream or this resource. However, I have an idea - which might seem a little complex.

How about we track each resource in a `PhantomReference`? When a `Resource` reference is enqueued, we can match up the reference with the input stream and force close it. We can't do `PhantomReference#get` to get the object being collected, but if we map this phantom reference to the input stream in question, we can ensure we close said stream.

It _might_ be overkill, but it is an option if we want to guard against plugins forgetting to close streams when they're done.

Reply:
That isn't a bad idea and should be explored...but that doesn't fix a "one and done" situation where a plugin asks for the stream, uses it, doesn't close it, never asked for again in the engine lifecycle.

Thoughts on that?

Reply:
It depends how they get the stream, but if we're encouraging the `Resource` to be the object that we try-with-resources around then if a plugin discards the resource (without closing), when the garbage collector kills the `Resource` the stream should be closed at that point regardless - so if a plugin grabs the stream from the resource this is still coming in to play. I would think that would solve the issue.

Ultimately, the suggestion is to tie the stream lifetime to the resource lifetime. If a plugin keeps hold of the `Resource`... well, really not much we can (or should) do about it, the plugin is being _really_ stupid then.

We should probably annotate `Resource` with `@DoNotStore` with that in mind.

Reply:
It is already annotated with `@DoNotStore`

Reply:
Oh... right, okay!